File: go/0340_longest_substring_with_at_most_K_distinct_characters.go
func lengthOfLongestSubstringKDistinct(s string, k int) int {
	mp := make(map[byte]int)
	res, left, right := 0, 0, 0

	slen := len(s)
	for right <= slen-1 {
		if mplen := len(mp); mplen <= k {
			if count, ok := mp[s[right]]; !ok {
				mp[s[right]] = 1
			} else {
				mp[s[right]] = count + 1
			}

			if len(mp) <= k && right-left+1 > res {
				res = right - left + 1
			}
			right++
		} else {
			count := mp[s[left]]
			if count == 1 {
				delete(mp, s[left])
			} else {
				mp[s[left]] = count - 1
			}
			left++
		}
	}
	return res
}

File: python/0340_longest_substring_with_at_most_k_distinct_characters.py
class Solution:
    def lengthOfLongestSubstringKDistinct(self, s, k):
        """
        :type s: str
        :type k: int
        :rtype: int
        """

        if k == 0 or len(s) == 0:
            return 0

        table = {}
        maxlength = 0

        left, right = 0, 0

        while right < len(s):
            table[s[right]] = table[s[right]] + 1 if s[right] in table else 1
            if len(table) <= k:
                if right - left + 1 > maxlength:
                    maxlength = right - left + 1
                right += 1
                continue

            while len(table) > k and left < right:
                table[s[left]] -= 1
                if table[s[left]] == 0:
                    del (table[s[left]])

                left += 1
            if len(table) <= k and right - left + 1 > maxlength:
                maxlength = right - left + 1
            right += 1
        return maxlength
